### ISSUE TITLE
Running bce loss is normal but Focal Loss and self define balanced cross entropy got problem

### ISSUE THREAD

**Issue by chinhong11** [ultralytics/ultralytics#5959]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

I was trying to run yolov8 with different loss function on my custom dataset, it was normal when I try with the default loss which is bce. When I change the loss to focal loss or balanced cross entropy, the results seems normal only until 140 epochs, it become very weird where my mAP50 and mAP50-95 was 0.000xx if I try to run 150 epochs and above but if it is 140 epochs, the mAP50 and mAP50-95 was 0.5xx.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
The issue you're describing seems to suggest that the model is not generalizing well when using alternative loss functions like Focal Loss or Balanced Cross Entropy, especially beyond 140 epochs. There could be a variety of reasons behind this issue, such as overfitting or unstable gradients. Below are some debugging steps you might want to consider:

#### Debugging Steps:
1. **Check Learning Rate**: Ensure the learning rate isn't too high, which could cause the model to converge to poor minima.
  
2. **Visualize Loss Curves**: Monitor the loss curves using tensorboard or similar tools to see if there are any anomalies like spikes or drastic changes.

3. **Model Inspection**: Use model inspection techniques like feature maps visualization to understand what the model is focusing on.

4. **Early Stopping**: Implement early stopping based on validation loss to prevent the model from overfitting.

5. **Data Augmentation**: Consider using more aggressive data augmentation techniques to improve generalization.

6. **Regularization**: Apply regularization techniques like dropout, if not already used.

7. **Check the Implementation**: Double-check your code for implementing Focal Loss and Balanced Cross Entropy to ensure there are no bugs.

#### Python Code to Monitor Loss using TensorBoard
Here's an example snippet to monitor training loss using TensorBoard. You may have something similar in your codebase.

```python
from torch.utils.tensorboard import SummaryWriter
from pathlib import Path

def log_loss_to_tensorboard(epoch, loss, log_dir=Path('runs/')):
    """
    Logs loss to TensorBoard.
    
    Args:
        epoch (int): Current epoch number.
        loss (float): Loss value to log.
        log_dir (Path): Path to the directory where TensorBoard logs are saved.
    """
    writer = SummaryWriter(log_dir)
    writer.add_scalar('Loss/train', loss, epoch)
    writer.close()
```

If you've checked all of these and are still encountering the issue, it might be worth reverting back to BCE loss to see if the issue persists. Then, incrementally switch to the new loss function while closely monitoring performance.

Hope this helps, and if the issue persists, please provide more details for further diagnosis.

**Comment by chinhong11** [comment]
@glenn-jocher Below is the code I implemented for Balanced Cross Entropy and I think it should be correct:

```class BCELossWithBalance(nn.Module):
    def __init__(self, weight=None, reduction='mean'):
        super(BCELossWithBalance, self).__init__()
        self.weight = weight
        self.reduction = reduction
    def forward(self, input, target):
        pos_weight = (target == 1).sum().float() / target.numel()
        weight = torch.full_like(target, 1)
        weight[target == 1] = pos_weight
        loss = nn.functional.binary_cross_entropy_with_logits(input, target, weight, reduction=self.reduction)
        return loss```
